In [1]:
import pandas as pd
import numpy as np

In [3]:
brands = pd.read_json('./brands.json', lines=True)

In [7]:
brands.describe(include='all')
# brands.describe()

_id       barcode category  \
count                                   1167  1.167000e+03     1012   
unique                                  1167           NaN       23   
top     {'$oid': '601ac115be37ce2ead437551'}           NaN   Baking   
freq                                       1           NaN      369   
mean                                     NaN  5.111115e+11      NaN   
std                                      NaN  2.874497e+05      NaN   
min                                      NaN  5.111110e+11      NaN   
25%                                      NaN  5.111112e+11      NaN   
50%                                      NaN  5.111114e+11      NaN   
75%                                      NaN  5.111117e+11      NaN   
max                                      NaN  5.111119e+11      NaN   

       categoryCode                                                cpg  \
count           517                                               1167   
unique           14                                                204   
top          BAKING  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
freq            359                                                 98   
mean            NaN                                                NaN   
std             NaN                                                NaN   
min             NaN                                                NaN   
25%             NaN                                                NaN   
50%             NaN                                                NaN   
75%             NaN                                                NaN   
max             NaN                                                NaN   

           name    topBrand brandCode  
count      1167  555.000000       933  
unique     1156         NaN       897  
top     Huggies         NaN            
freq          2         NaN        35  
mean        NaN    0.055856       NaN  
std         NaN    0.229850       NaN  
min         NaN    0.000000       NaN  
25%         NaN    0.000000       NaN  
50%         NaN    0.000000       NaN  
75%         NaN    0.000000       NaN  
max         NaN    1.000000       NaN

`id` is all unique, which is good.

In [15]:
pd.crosstab(brands.category, brands.categoryCode, dropna=False)

categoryCode                 BABY  BAKING  BEER_WINE_SPIRITS  BEVERAGES  \
category                                                                  
Baby                            7       0                  0          0   
Baking                          0     359                  0          0   
Beer Wine Spirits               0       0                 31          0   
Beverages                       0       0                  0          1   
Bread & Bakery                  0       0                  0          0   
Candy & Sweets                  0       0                  0          0   
Cleaning & Home Improvement     0       0                  0          0   
Dairy & Refrigerated            0       0                  0          0   
Frozen                          0       0                  0          0   
Grocery                         0       0                  0          0   
Health & Wellness               0       0                  0          0   
Magazines                       0       0                  0          0   
Outdoor                         0       0                  0          0   
Personal Care                   0       0                  0          0   

categoryCode                 BREAD_AND_BAKERY  CANDY_AND_SWEETS  \
category                                                          
Baby                                        0                 0   
Baking                                      0                 0   
Beer Wine Spirits                           0                 0   
Beverages                                   0                 0   
Bread & Bakery                              5                 0   
Candy & Sweets                              0                71   
Cleaning & Home Improvement                 0                 0   
Dairy & Refrigerated                        0                 0   
Frozen                                      0                 0   
Grocery                                     0                 0   
Health & Wellness                           0                 0   
Magazines                                   0                 0   
Outdoor                                     0                 0   
Personal Care                               0                 0   

categoryCode                 CLEANING_AND_HOME_IMPROVEMENT  \
category                                                     
Baby                                                     0   
Baking                                                   0   
Beer Wine Spirits                                        0   
Beverages                                                0   
Bread & Bakery                                           0   
Candy & Sweets                                           0   
Cleaning & Home Improvement                              6   
Dairy & Refrigerated                                     0   
Frozen                                                   0   
Grocery                                                  0   
Health & Wellness                                        0   
Magazines                                                0   
Outdoor                                                  0   
Personal Care                                            0   

categoryCode                 DAIRY_AND_REFRIGERATED  FROZEN  GROCERY  \
category                                                               
Baby                                              0       0        0   
Baking                                            0       0        0   
Beer Wine Spirits                                 0       0        0   
Beverages                                         0       0        0   
Bread & Bakery                                    0       0        0   
Candy & Sweets                                    0       0        0   
Cleaning & Home Improvement                       0       0        0   
Dairy & Refrigerated                              5       0        0   
Frozen       

There seems to be a misalignment with the `category` column and the `categoryCode` column. While the alignment for 14 categories is correct, as indicated by the diagonal matrix above, there are still 9 categories in the `category` column that are unaccounted for.

In [20]:
brands.topBrand.isna().sum()

612

There seems to be an issue with the `topBrand` column. There are 612 NA values in a True/False column. We would need to decide if imputation is a viable option. Otherwise, we need to sort that out in the data processing step.

In [25]:
brands.groupby(['brandCode']).count()

_id  barcode  category  categoryCode  cpg  name  topBrand
brandCode                                                                    
                     35       35        35             0   35    35        32
 MILANO               1        1         1             0    1     1         1
 SOBE                 1        1         1             0    1     1         0
 TAZO BOTTLED TEAS    1        1         1             0    1     1         0
09090909090           1        1         1             0    1     1         1
...                 ...      ...       ...           ...  ...   ...       ...
YOPLAIT GO-GURT       1        1         1             0    1     1         0
YQ YOPLAIT            1        1         1             0    1     1         0
YUBAN                 1        1         1             0    1     1         1
ZIMA                  1        1         1             0    1     1         0
ZUMBIDA               1        1         1             0    1     1         0

[897 rows x 7 columns]

In [28]:
brands[brands.brandCode == ''].head(10)

_id       barcode             category  \
124  {'$oid': '57ebc11fe4b0ac389136a33a'}  511111802075               Baking   
153  {'$oid': '58861c7d4e8d0d20bc42c4d6'}  511111601449               Snacks   
163  {'$oid': '57ebc2ace4b0ac389136a346'}  511111801962                 Deli   
188  {'$oid': '58b59989e4b0857c2ddb7255'}  511111400998    Beer Wine Spirits   
234  {'$oid': '58b5988ce4b0857c2ddb7252'}  511111301028    Beer Wine Spirits   
236  {'$oid': '57ebc125e4b0ac389136a33b'}  511111302063              Grocery   
278  {'$oid': '585a967fe4b03e62d1ce0e80'}  511111801689               Snacks   
288  {'$oid': '585a96d2e4b03e62d1ce0e89'}  511111001607            Beverages   
297  {'$oid': '580e015be4b0f32b2de21385'}  511111501824  Condiments & Sauces   
302  {'$oid': '588b9ff4e4b02187f85cdadb'}  511111901099  Condiments & Sauces   

    categoryCode                                                cpg  \
124          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
153          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
163          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
188          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f709e4b0...   
234          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f709e4b0...   
236          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
278          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
288          NaN  {'$ref': 'Cogs', '$id': {'$oid': '5332f5fbe4b0...   
297          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   
302          NaN  {'$ref': 'Cogs', '$id': {'$oid': '559c2234e4b0...   

                                      name  topBrand brandCode  
124                         Kraft Caramels       0.0            
153  Jell-O Refrigerated Pudding & Gelatin       0.0            
163                                     P3       0.0            
188                          Redd's Wicked       0.0            
234                 Henry's Hard Sparkling       0.0            
236                Kraft Macaroni & Cheese       0.0            
278                    Lay's Kettle Cooked       1.0            
288                             Pepsi Diet       0.0            
297                        Kraft BBQ Sauce       0.0            
302                   Kraft Salad Dressing       0.0

There also seems to be an issue with the `brandCode` column. There is a blank value that is being registered for 35 products, some of which do have associated brand codes.